crates/goose/src/session/claude_code.rs

```diff
@@ -0,0 +1,273 @@
+
+    tracing::debug!("Checking for Claude Code sessions in: {:?}", projects_dir);
+
+    if !projects_dir.exists() {
+        tracing::debug!("Claude projects directory does not exist");
+        return Ok(Vec::new());
+    }
+
+    let mut sessions = Vec::new();
+
+    for entry in std::fs::read_dir(projects_dir)? {
+        let entry = entry?;
+        let project_path = entry.path();
+
+        if !project_path.is_dir() {
+            continue;
+        }
+
+        for file_entry in std::fs::read_dir(&project_path)? {
+            let file_entry = file_entry?;
+            let file_path = file_entry.path();
+
+            if !file_path.is_file() {
+                continue;
+            }
+
+            let file_name = file_path.file_name().and_then(|n| n.to_str()).unwrap_or("");
+
+            if file_name.starts_with("agent-") || !file_name.ends_with(".jsonl") {
+                continue;
+            }
+
+            if let Ok((session_id, working_dir, updated_at)) = parse_session_metadata(&file_path) {
+                tracing::debug!(
+                    "Found Claude session: {} updated at {}",
+                    session_id,
+                    updated_at
+                );
+                sessions.push((session_id, working_dir, updated_at));
+            }
+        }
+    }
+
+    tracing::debug!("Total Claude sessions found: {}", sessions.len());
+
+    sessions.sort_by(|a, b| b.2.cmp(&a.2));
+    sessions.truncate(10);
+
+    Ok(sessions)
+}
+
+fn parse_session_metadata(file_path: &PathBuf) -> Result<(String, PathBuf, DateTime<Utc>)> {
+    let file = File::open(file_path)?;
+    let reader = BufReader::new(file);
+
+    let mut session_id = None;
+    let mut working_dir = None;
+    let mut latest_timestamp = None;
+
+    for line in reader.lines() {
+        let line = line?;
+        if let Ok(entry) = serde_json::from_str::<ClaudeJsonLine>(&line) {
+            if session_id.is_none() {
+                session_id = entry.session_id;
+            }
+            if working_dir.is_none() {
+                working_dir = entry.cwd;
+            }
+            if let Some(ts) = entry.timestamp {
+                if let Ok(dt) = DateTime::parse_from_rfc3339(&ts) {
+                    latest_timestamp = Some(dt.with_timezone(&Utc));
+                }
+            }
+        }
+    }
+
+    let session_id = session_id.ok_or_else(|| anyhow::anyhow!("No session ID found"))?;
+    let working_dir = working_dir.ok_or_else(|| anyhow::anyhow!("No working dir found"))?;
+    let updated_at = latest_timestamp.unwrap_or_else(Utc::now);
+
+    Ok((session_id, PathBuf::from(working_dir), updated_at))
+}
+
+pub fn load_claude_code_session(session_id: &str) -> Result<Conversation> {
+    let home = dirs::home_dir().ok_or_else(|| anyhow::anyhow!("No home dir"))?;
+    let projects_dir = home.join(".claude").join("projects");
+
+    for entry in std::fs::read_dir(projects_dir)? {
```

Comment:
this code is the same as on 41; factor our either the iteration or do the dumb thing and fish out the session out of the list